File: 2024/Day 06/Python/main.py
"""Advent of Code - 2024 - Day 6"""
from itertools import pairwise

type State = tuple[tuple[int, int], int]
type Map = set[tuple[int, int]]

dirs = [(0, -1), (1, 0), (0, 1), (-1, 0)]

def traverse(initial_state: State, obstacles: Map, width: int, height: int):
    """simulates the guard traversing the map"""
    (guard_pos, turns) = initial_state
    while 0 <= guard_pos[0] < width and 0 <= guard_pos[1] < height:
        yield (guard_pos, turns)
        guard_dir = dirs[turns]
        next_guard_pos = (guard_dir[0] + guard_pos[0], guard_dir[1] + guard_pos[1])
        if next_guard_pos in obstacles:
            turns = (turns + 1) % 4
        else:
            guard_pos = next_guard_pos

def get_obstacle_pos(initial_state: State, obstacles: Map, width: int, height: int):
    """yields candidate obstacle positions"""
    checked: set[tuple[int, int]] = set()
    for new_start, (obstacle_pos, _) in pairwise(traverse(initial_state, obstacles, width, height)):
        if obstacle_pos not in checked:
            checked.add(obstacle_pos)
            yield (new_start, obstacle_pos)

def detect_loop(initial_state: State, obstacles: Map, width: int, height: int):
    """detects if the guard would make a loop"""
    state_set: set[tuple[tuple[int, int], int]] = set()
    for state in traverse(initial_state, obstacles, width, height):
        if state in state_set:
            return True
        state_set.add(state)
    return False

if __name__ == '__main__':
    with open(0, encoding="utf-8") as f:
        lines: list[str] = f.read().splitlines()
    i_obstacles: Map = set()
    i_state: State = ((-1, -1), -1)
    w: int = len(lines[0])
    h: int = len(lines)
    for (y, line) in enumerate(lines):
        for (x, char) in enumerate(line):
            match char:
                case "#":
                    i_obstacles.add((x, y))
                case "^":
                    i_state = ((x, y), 0)
                case ">":
                    i_state = ((x, y), 1)
                case "v":
                    i_state = ((x, y), 2)
                case "<":
                    i_state = ((x, y), 3)
                case _:
                    pass

    print((len({ pos for (pos, _) in traverse(i_state, i_obstacles, w, h) })))
    loops: int = 0
    for start, obs_pos in get_obstacle_pos(i_state, i_obstacles, w, h):
        if detect_loop(start, i_obstacles | {obs_pos}, w, h):
            loops += 1
    print(loops)

File: 2024/Day 06/C++/main.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_set>

std::pair<int, int> get_dir(int turns)
{
    std::pair<int, int> p;
    if (turns == 0)
    {
        p.first = 0;
        p.second = -1;
    }
    else if (turns == 1)
    {
        p.first = 1;
        p.second = 0;
    }
    else if (turns == 2)
    {
        p.first = 0;
        p.second = 1;
    }
    else if (turns == 3)
    {
        p.first = -1;
        p.second = 0;
    }
    return p;
}

int coord_to_int(int x, int y)
{
    return x * 256 + y;
}

int coord_to_int(std::pair<int, int> pos)
{
    return coord_to_int(pos.first, pos.second);
}

bool in_bounds(std::pair<int, int> &pos, int w, int h)
{
    return 0 <= pos.first && pos.first < w && 0 <= pos.second && pos.second < h;
}

std::pair<std::pair<int, int>, int> step(std::pair<std::pair<int, int>, int> &state, std::unordered_set<int> &obstacles)
{
    std::pair<int, int> guard_pos = state.first;
    int turns = state.second;
    std::pair<int, int> guard_dir = get_dir(turns);
    std::pair<int, int> next_guard_pos;
    next_guard_pos.first = guard_pos.first + guard_dir.first;
    next_guard_pos.second = guard_pos.second + guard_dir.second;
    if (obstacles.find(coord_to_int(next_guard_pos)) != obstacles.end())
    {
        state.second = (turns + 1) % 4;
    }
    else
    {
        state.first = next_guard_pos;
    }
    return state;
}

int part1(std::pair<std::pair<int, int>, int> state, std::unordered_set<int> obstacles, int w, int h)
{
    std::unordered_set<int> traversed;
    while (in_bounds(state.first, w, h))
    {
        traversed.insert(coord_to_int(state.first));
        state = step(state, obstacles);
    }
    return traversed.size();
}

int main(int argc, char *argv[])
{
    std::cout << "start" << std::endl;
    std::string line;
    int y = 0;
    auto hash = [](const std::pair<int, int> &p)
    { return p.first * 256 + p.second; };
    std::unordered_set<int> obstacles;
    std::pair<std::pair<int, int>, int> state;
    while (std::getline(std::cin, line))
    {
        for (int x = 0; x < line.length(); x++)
        {
            if (line[x] == '#')
            {
                obstacles.insert(coord_to_int(x, y));
            }
            else if (line[x] == '^')
            {
                state = std::make_pair(std::make_pair(x, y), 0);
            }
            else if (line[x] == '>')
            {
                state = std::make_pair(std::make_pair(x, y), 1);
            }
            else if (line[x] == 'v')
            {
                state = std::make_pair(std::make_pair(x, y), 2);
            }
            else if (line[x] == '<')
            {
                state = std::make_pair(std::make_pair(x, y), 3);
            }
        }
        y++;
    }
    int h = y;
    int w = line.length();
    std::cout << part1(state, obstacles, h, w) << std::endl;
}